### ISSUE TITLE
conform and umbrella: *.conf file is not copied

### ISSUE THREAD

**Issue by ahamez** [bitwalker/distillery#56]
### Steps to reproduce

When using distillery with conform and an umbrella project, the *.conf file of sub apps is not copied to release folder.
### Verbose Logs

```
$ MIX_ENV=prod mix release --verbose --env=prod
==> Loading configuration..
==> Assembling release..
==> Building release bar:0.1.0 using environment prod
==> Discovered applications:
  foo-0.1.0
    from: _build/prod/lib/foo
    applications:
      :kernel
      :stdlib
      :elixir
      :conform
      :logger
    includes: none

  sasl-3.0
    from: /usr/local/Cellar/erlang/19.0.2/lib/erlang/lib/sasl-3.0
    applications:
      :kernel
      :stdlib
    includes: none

  iex-1.3.2
    from: /usr/local/Cellar/elixir/1.3.2/bin/../lib/iex
    applications:
      :kernel
      :stdlib
      :elixir
    includes: none

  elixir-1.3.2
    from: /usr/local/Cellar/elixir/1.3.2/bin/../lib/elixir
    applications:
      :kernel
      :stdlib
      :compiler
    includes: none

  kernel-5.0
    from: /usr/local/Cellar/erlang/19.0.2/lib/erlang/lib/kernel-5.0
    applications: none
    includes: none

  stdlib-3.0.1
    from: /usr/local/Cellar/erlang/19.0.2/lib/erlang/lib/stdlib-3.0.1
    applications:
      :kernel
    includes: none

  compiler-7.0.1
    from: /usr/local/Cellar/erlang/19.0.2/lib/erlang/lib/compiler-7.0.1
    applications:
      :kernel
      :stdlib
    includes: none

  conform-2.1.1
    from: _build/prod/lib/conform
    applications:
      :kernel
      :stdlib
      :elixir
      :neotoma
    includes: none

  neotoma-1.7.3
    from: _build/prod/lib/neotoma
    applications:
      :kernel
      :stdlib
    includes: none

  logger-1.3.2
    from: /usr/local/Cellar/elixir/1.3.2/bin/../lib/logger
    applications:
      :kernel
      :stdlib
      :elixir
    includes: none

==> conform: loading schema
==> conform: no schema found, skipping
==> Generated overlay vars:
    release_name=:bar
    release_version="0.1.0"
    is_upgrade=false
    upgrade_from=:latest
    dev_mode=false
    include_erts=true
    include_src=false
    include_system_libs=true
    erl_opts=""
    erts_vsn="8.0.2"
    output_dir="rel/bar"
==> Copying applications to rel/bar
==> Generating nodetool
==> Generating start_erl.data
==> Generating vm.args
==> Generating sys.config from config/config.exs
==> Including ERTS 8.0.2 from /usr/local/Cellar/erlang/19.0.2/lib/erlang/erts-8.0.2
==> Generating boot script
==> Generating RELEASES
==> Generating start_clean.boot
==> Applying overlays
==> Applying mkdir overlay
    dst: releases/0.1.0/hooks
==> Applying copy overlay
    src: _build/prod/lib/conform/priv/bin/pre_start.sh
    dst: releases/0.1.0/hooks/pre_start
==> Applying mkdir overlay
    dst: releases/0.1.0/commands
==> Packaging release..
==> Archiving bar-0.1.0
==> Writing tarball to rel/bar/releases/0.1.0/bar.tar.gz
==> Updating tarball
==> Tarball updated!
==> Release successfully built!
    You can run it in one of the following ways:
      Interactive: rel/bar/bin/bar console
      Foreground: rel/bar/bin/bar foreground
      Daemon: rel/bar/bin/bar start
```
### Description of issue
- What are the expected results?
  The *.conf file of the subapp using conform should be copied to the release folder.
- What OS, Erlang/Elixir versions are you seeing this issue on?
  Mac OS X 10.11.6, Erlang/OTP 19 [erts-8.0.2], Elixir (1.3.2)
- If possible, also provide your `rel/config.exs`, as it is often
  my first troubleshooting question, and you'll save us both time :)
  I created a small project here to reproduce the problem: https://github.com/ahamez/distillery-umbrella-conform

In the foo sub app, there is a config/foo.schema.exs, but logs of distillery seem to say otherwise. Maybe schemas should be in a different place when using distillery?

**Comment by martin-langhoff** [comment]
I seem to be affected by the same bug.` foo.schema.exs` is also missing in the release.

**Comment by coryodaniel** [comment]
Hey guys, I see that all the tickets around this are merged and closed. I'm using the master branches of distillery and conform and having the same issues as outlined above. Is there anything that need to be changed in the configuration to get this to work? I set up a demo app of the [failure](https://github.com/coryodaniel/distillery_umbrella_conform). I also made some [comments](https://github.com/bitwalker/conform/issues/91#issuecomment-257181715) on another ticket that seems to be regarding the same issue.

Here is the output of `mix release`

```bash
$ MIX_ENV=prod mix release --verbose --env=prod
==> Loading configuration..
==> Assembling release..
==> Building release distillery_umbrella_conform:0.1.0 using environment prod
==> One or more direct or transitive dependencies are missing from
    :applications or :included_applications, they will not be included
    in the release:

    :conform

    This can cause your application to fail at runtime. If you are sure
    that this is not an issue, you may ignore this warning.

==> Discovered applications:
  logger-1.3.4
    from: /usr/local/Cellar/elixir/1.3.4/bin/../lib/logger
    applications:
      :kernel
      :stdlib
      :elixir
    includes: none

  compiler-7.0.2
    from: /usr/local/Cellar/erlang/19.1/lib/erlang/lib/compiler-7.0.2
    applications:
      :kernel
      :stdlib
    includes: none

  stdlib-3.1
    from: /usr/local/Cellar/erlang/19.1/lib/erlang/lib/stdlib-3.1
    applications:
      :kernel
    includes: none

  kernel-5.1
    from: /usr/local/Cellar/erlang/19.1/lib/erlang/lib/kernel-5.1
    applications: none
    includes: none

  elixir-1.3.4
    from: /usr/local/Cellar/elixir/1.3.4/bin/../lib/elixir
    applications:
      :kernel
      :stdlib
      :compiler
    includes: none

  iex-1.3.4
    from: /usr/local/Cellar/elixir/1.3.4/bin/../lib/iex
    applications:
      :kernel
      :stdlib
      :elixir
    includes: none

  sasl-3.0.1
    from: /usr/local/Cellar/erlang/19.1/lib/erlang/lib/sasl-3.0.1
    applications:
      :kernel
      :stdlib
    includes: none

  app_a-0.1.0
    from: _build/prod/lib/app_a
    applications:
      :kernel
      :stdlib
      :elixir
      :logger
    includes: none

==> conform: loading schema
==> conform: no schema found, skipping
==> Generated overlay vars:
    release_name=:distillery_umbrella_conform
    release_version="0.1.0"
    is_upgrade=false
    upgrade_from=:latest
    dev_mode=false
    include_erts=true
    include_src=false
    include_system_libs=true
    erl_opts=""
    erts_vsn="8.1"
    output_dir="rel/distillery_umbrella_conform"
==> Copying applications to rel/distillery_umbrella_conform
==> Generating nodetool
==> Generating start_erl.data
==> Generating vm.args
==> Generating sys.config from config/config.exs
==> Including ERTS 8.1 from /usr/local/Cellar/erlang/19.1/lib/erlang/erts-8.1
==> Generating boot script
==> Generating RELEASES
==> Generating start_clean.boot
==> Applying overlays
==> Applying mkdir overlay
    dst: releases/0.1.0/hooks
==> Applying mkdir overlay
    dst: releases/0.1.0/hooks/pre_start.d
==> Applying mkdir overlay
    dst: releases/0.1.0/hooks/post_start.d
==> Applying mkdir overlay
    dst: releases/0.1.0/hooks/pre_stop.d
==> Applying mkdir overlay
    dst: releases/0.1.0/hooks/post_stop.d
==> Applying copy overlay
    src: _build/prod/lib/conform/priv/bin/pre_start.sh
    dst: releases/0.1.0/hooks/pre_start.d/00_pre_start_hook.sh
==> Applying mkdir overlay
    dst: releases/0.1.0/commands
==> Packaging release..
==> Archiving distillery_umbrella_conform-0.1.0
==> Writing tarball to rel/distillery_umbrella_conform/releases/0.1.0/distillery_umbrella_conform.tar.gz
==> Updating tarball
==> Tarball updated!
==> Release successfully built!
```

Note the:
```
==> conform: loading schema
==> conform: no schema found, skipping
```

Ive had great luck w/ both these tools for a non-umbrella app. Thanks for all the hard work and any help/direction :D

**Comment by martin-langhoff** [comment]
Right now`master` is a bit out of sync between distillery and conform. The hooks API is a bit ahead in distillery, and the matching changes are pending in conform. 

Perhaps try the latest tagged of each, I think that combo is working reasonably well (I'm in the midst of testing exactly that :-) ).

**Comment by coryodaniel** [comment]
@martin-langhoff I was using the hex versions that are the same version numbers as the newest tags, I just switched to using the tagged versions in case they were different (tags 0.10.1, 2.1.1). Still seeing the same thing? Is there a different branch/tag I should be looking for?

**Comment by martin-langhoff** [comment]
@coryodaniel your opening post mentioned being on `master`. AFAICS this stuff works, you might need to try seeing whether you can repro your bug, perhaps from the `exrm-umbrella-test` project.

For example, I've used this... https://github.com/martin-langhoff/exrm-umbrella-test/tree/more-interesting.

**Comment by coryodaniel** [comment]
I was originally making comments on another ticket where someone was having the same issues, I tried both masters, both latest tags, and both releases in hex.

Here is a demo app w/ the failure -> https://github.com/coryodaniel/distillery_umbrella_conform

Here is a screenshot of the release folder, missing the *conf files.
<img width="333" alt="66ffceda-9eaf-11e6-9adb-e57d6c23b321" src="https://cloud.githubusercontent.com/assets/11587/19866442/60a3263c-9f5d-11e6-88e7-dad8e46f97ca.png">

**Comment by coryodaniel** [comment]
So I ran this branch https://github.com/martin-langhoff/exrm-umbrella-test/tree/more-interesting and got the following...

A config file in the 'master_app'
<img width="556" alt="screen shot 2016-10-31 at 5 01 25 pm" src="https://cloud.githubusercontent.com/assets/11587/19875661/d996c8c2-9f8b-11e6-8df4-8494d95fddbc.png">

but it looked like that branch wasn't configure for doing an umbrella, so I added an umbrella release to rel/config.exs

```elixir
# appended...
release :umbrella_test do
  set pre_start_hook: conform_prestart
  plugin Conform.ReleasePlugin
  set version: "0.1.0"
  set applications: [
    master_app: :permanent
  ]
end
```

And when I build the release ended up with an umbrella app, but without a config file (also, its in the _build directory instead of rel/)
<img width="574" alt="screen shot 2016-10-31 at 5 01 42 pm" src="https://cloud.githubusercontent.com/assets/11587/19875684/1227772c-9f8c-11e6-9a64-af74dffa4108.png">

**Comment by bitwalker** [comment]
I've been working on some changes today:

- In distillery master, release output is now in _build
- In conform master, it now uses overlays + the new `pre_start.d` hooks directory,
  so you no longer need to set the `pre_start_hook` config setting in your release config.

I can start looking into this more, but I've been trying to clean up a few other pending tasks. @martin-langhoff is your exrm-umbrella-test project able to reproduce this, or should I work with the example that @coryodaniel put together?

**Comment by coryodaniel** [comment]
@bitwalker let me know if there is anything I can help with. Really dig both of the bullet points above! 👍 👍

**Comment by coryodaniel** [comment]
@bitwalker the 'more-interesting' branch on @martin-langhoff's repo has been ported to distillery if you want to use it. The only thing I had to add to it was a release in `rel/config.exs` for the umbrella. It currently only has releases for the individual apps.

My app is boring and only has one app in the umbrella, I can make a less trivial example if that helps (not sure if more complicated anything has ever been helpful :D)

**Comment by bitwalker** [comment]
So I've updated `exrm-umbrella-test` (it is now `distillery-umbrella-test`) with @martin-langhoff's changes, and updates to reflect conform master, as well as fixed a bug I found in Distillery's handling of copying apps (unrelated to this issue though). I've determined the source of the issue, which is that if the name of the release does not match the name of an app in the umbrella, no config will be found. I'm a little unsure of what the right thing to do is when defining a release with child apps, all of which have a schema - which schema is the right one? Is it implicitly a blank schema with `extends: [..list of apps in umbrella..]`? In any case, in the example, `master_app` actually includes the other children, so building a release of `master_app` is implicitly building a release of the whole umbrella, but in that case we have an "entry" schema which we can consider the parent. I haven't considered that we could have top-level sibling schemas. Thoughts?

Also, could one or both of you give the updated example app a shot? 

My repro steps are as follows:

- Clone https://github.com/bitwalker/distillery-umbrella-test
- `mix do deps.get, compile`
- `mix release --profile=master_app:prod`
- `_build/dev/rel/master_app/bin/master_app console`
- `Application.get_all_env(:app_one)` should return 8090 instead of 9090 for the port, this will only occur if conform is working and it has included the schemas from the child apps.

**Comment by coryodaniel** [comment]
Ran the example and got 8090 as `:app_one`'s port. In that release `--profile=master_app:prod` we are only releasing a single app (with library app dependencies) in the umbrella though correct - not the whole umbrella "UmbrellaTest"? Just want to make sure I'm following.

So funny thing, when I noticed this wasn't working my first thought was that I probably had to add a schema and extend all of my umbrella app's schemas. So I added a .schema.exs and a .conf file to releases/VSN/. My schema was blank except extends, my .conf was all my conf files concatenated. I wasn't sure what to do about a .schema.ez file though. Running the console outputted the message:

`
$ _build/prod/rel/umbrella/bin/umbrella console
escript: Failed to open file: /Users/coryodaniel/Workspace/umbrella/_build/prod/rel/umbrella/releases/0.1.0/conform
`

I think a blank schema at the umbrella level would be fine...

Just some info, our umbrella app is:
- biz models
- https:// phoenix app (OTP)
- wss:// phoenix app (OTP)
- some other non-OTP apps

I could see a case where our containers were running the umbrella as a whole while other contains may only be running a portion of it.

**Comment by bitwalker** [comment]
> In that release --profile=master_app:prod we are only releasing a single app (with library app dependencies) in the umbrella though correct - not the whole umbrella "UmbrellaTest"? Just want to make sure I'm following.

Well in the example project, `master_app`, `app_one`, and `app_two` are all fully supervised OTP applications, `master_app` depends on `app_one` and `app_two`, so building a release of `master_app` effectively builds a release of the entire umbrella. That's obviously only true of the example app though, not every umbrella will have such a structure, it just happens to be in this case.

> I wasn't sure what to do about a .schema.ez file though

The schema.ez file contains the extended schemas, and imported modules, but the primary issue is that currently if `conform` doesn't detect a schema for the current release, it doesn't bundle itself into the release, and so you're missing the `conform` binary (as you saw).

So basically, it looks like we probably want conform to check if the current release is an umbrella, if so, generate a default schema which imports the schemas of all child applications, concatenates the .conf files into a single one, and generally does what it needs to do in order to work in this configuration. I'll try to play around with that tomorrow.

**Comment by coryodaniel** [comment]
@bitwalker sounds like a good approach. I'll have some time to give it a swing on Saturday if you don't have time to get around to it

**Comment by coryodaniel** [comment]
Hey, this weekend got away from me. Will def give a swing at this tomorrow!

**Comment by coryodaniel** [comment]
So, I'm assuming the issue is actually in the conform plugin. I haven't quiet figured it out, but I am still seeing this is in the log when you are building an umbrella.

```
conform: loading schema
conform: no schema found, skipping
```

Im trying to figure out how how to tell at the plugin level if the `release` received by `before_assembly/1` is an umbrella or not... I'm assuming I'll have to put that into the struct from the distillery side?

**Comment by coryodaniel** [comment]
So in an [umbrella](https://github.com/coryodaniel/distillery-umbrella-test/commit/d3f935c1f5e15c01b38ff630588418cbbb8b4cef) in the release when you set how the applications should be started (below) if you are listing your one OTP app (master_app) I'm assuming the generated schema and conf should still be all of your apps, right? app_one could have a configuration that needs to be included as well... Is there a more reliable way of finding the applications under the umbrella than globbing the `apps/` directory?

```elixir
release :umbrella_test do
  plugin Conform.ReleasePlugin
  set pre_start_hook: conform_prestart
  set version: "0.1.0"
  set applications: [
    master_app: :permanent
  ]
end
```

**Comment by coryodaniel** [comment]
Im assuming since we are concat and merging the conf and schema from conf and schema files people are already editing, the actual conf/schema files for the umbrella should probably not be actual files in the repo, probably just made in memory and spit out in the release folder. @bitwalker, thoughts?

**Comment by coryodaniel** [comment]
I think this might be entirely related to conform's plugin. Here is a link to my [wip](https://github.com/bitwalker/conform/pull/94/files). Should I start a ticket in that repo and link to this one or should we keep this open in case there is distillery related changes?

**Comment by bitwalker** [comment]
Let's close this one for now and use the PR thread for further discussion on this one. We can reopen this if there is work in distillery which needs to happen.

**Comment by coryodaniel** [comment]
If anyone on this ticket is interested, it looks like I have a fix in Conform for this -> https://github.com/bitwalker/conform/pull/94
